proxy_manager.go

```diff
@@ -142,24 +156,37 @@ func (m *proxyManager) remove(serviceName string, objectName string) bool {
 	return true
 }
 
-func (m *proxyManager) proxyFor(ctx context.Context, serviceName string, objectName string) (*proxy, error) {
+func (m *proxyManager) proxyWrapperFor(
+	ctx context.Context,
+	serviceName string,
+	objectName string,
+	wrapProxyFn func(p *proxy) (interface{}, error)) (interface{}, error) {
+
 	name := makeProxyName(serviceName, objectName)
 	m.mu.RLock()
-	obj, ok := m.proxies[name]
+	wrapper, ok := m.proxies[name]
 	m.mu.RUnlock()
 	if ok {
-		return obj, nil
+		return wrapper, nil
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
```

Comment:
I see it's still WIP but an early comment: Here, deferring the unlock holds the writer lock until the member responses back. This will block readers especially when - say, there is a retry or slow operation happens for this call. Do you think that makes sense to consider or it is negligible? I mean, all operations are supposed to be slow under such a condition?

Reply:
> EDIT: never mind, I missed the assignment at the very end. The comment does not make sense.

Good to know. I think, we were blocking all callers while the network I/O was happening here before my PR. In theory, we could improve that with striped locking, but I have a feeling that such optimization would only help in edge scenarios. WDYT?

Reply:
> we could improve that with striped locking, but I have a feeling that such optimization would only help in edge scenarios.

Yes I also was wondering how likely such edge cases are to happen. But I feel like it's not worth considering - at least for this block.

Reply:
Such edge cases assume that you're doing crazy things, say, like creating a new data structure on each HTTP request, or something that weird.